marpa/lib/Guacamole.pm

```diff
@@ -295,6 +295,23 @@ NonLiteral ::= Variable
              | PackageArrow
              | ParenExpr ElemSeq0
              | OpNullaryKeywordExpr
+             | DiamondExpr
+
+DiamondExpr ::= Diamond
+              | DoubleDiamond
+
+# This is written this way because of whitespace rules
+Diamond ~ '<' '$' VarIdent '>'
+        | '<>'
+
+# This is written this way because of whitespace rules
+DoubleDiamond ~ '<<' '$' VarIdent '>>'
+              | '<<>>'
+
+# This is here only for the Diamond/DoubleDiamond
+# because using VarName would allow a space
+# bah...
+VarIdent ~ [a-zA-Z0-9_]+
```

Comment:
We should probably fix `VarName` instead, as far as I understand spaces are not allowed inside variable names, not even in complex paths.

Reply:
`$ foo` is technically fine in Perl. I don't think we should allow it though. Do you think this fix needs to be done on this commit or as part of an overhaul of the idents and namings?

There's:
* Idents for calling a subroutine
* Idents for naming a subroutine
* Idents for variable names

I think I eventually merged the first two. This prevented `Foo::` from working (since you can't name a subroutine that, you couldn't make a call to a subroutine this way), which I'm actually fine with.

Then again, using semicolons anywhere within the idents (for both subroutines and variables) is fine. We currently recognize double colon as a separator between names but what about `::foo()` or `::$foo`?

The difference between the first two and the last one, as far as I can tell is:

* You cannot use QLike-value when calling subroutines or when naming them. (`q ()` fails, and `sub q{}` will fail.)

Am I missing anything else?

Reply:
This more complicated than I thought when I wrote the comment. If you think an overhaul is in order, then of course we should do it in a separate PR. :)